Sources/TuistGenerator/Mappers/AutogeneratedSchemesProjectMapper.swift

```diff
@@ -1,49 +1,189 @@
+            } else if let watch2ExtensionTarget = buildTargets.first(where: { $0.product == .watch2Extension }) {
+                actualRunTarget = hostWatchAppTargetReference(for: watch2ExtensionTarget, project: project)
+            } else {
+                actualRunTarget = nil
+            }
+        }
+
+        let runAction = actualRunTarget.map {
+            RunAction(
+                configurationName: buildConfiguration,
+                attachDebugger: true,
+                preActions: [],
+                postActions: [],
+                executable: .init(projectPath: project.path, name: $0.name),
+                filePath: nil,
+                arguments: defaultArguments(for: $0),
+                diagnosticsOptions: [.mainThreadChecker]
+            )
+        }
 
-        let buildTargets = buildableTargets(targetReference: targetReference, target: target, project: project)
-        let testTargets = testableTargets(targetReference: targetReference, target: target, project: project)
-        let executable = runnableExecutable(targetReference: targetReference, target: target, project: project)
+        // build targets should be first, because in case of extension schemes, the extension target should be the first one
+        var buildTargetReferences = (
+            buildTargets.sorted { $0.name < $1.name } + runTargets.sorted { $0.name < $1.name }
+        )
+        .map {
+            TargetReference(projectPath: project.path, name: $0.name)
+        }
+        if let runActionTargetReference = runAction?.executable,
+           !buildTargetReferences.contains(runActionTargetReference)
+        {
+            buildTargetReferences.append(runActionTargetReference)
+        }
 
-        let arguments = defaultArguments(for: target)
+        let testableTargets: [TestableTarget] = testTargets
+            .map {
+                let parallel = project.options.testingOptions.contains(.parallelizable)
+                let randomExecution = project.options.testingOptions.contains(.randomExecutionOrdering)
+                return TestableTarget(
+                    target: .init(projectPath: project.path, name: $0.name),
+                    parallelizable: parallel,
+                    randomExecutionOrdering: randomExecution
+                )
+            }
+            .sorted { $0.target.name < $1.target.name }
 
+        let buildAction: BuildAction
+        if buildTargetReferences.isEmpty {
+            buildAction = .init(targets: testableTargets.map(\.target))
+        } else {
+            buildAction = .init(targets: buildTargetReferences)
+        }
         return Scheme(
-            name: target.name,
+            name: name,
             shared: true,
-            buildAction: BuildAction(targets: buildTargets),
-            testAction: TestAction(
-                targets: testTargets,
+            buildAction: buildAction,
+            testAction: testableTargets.isEmpty ? nil : TestAction(
```

Comment:
Seems Xcode doesn't like a non existent `TestAction`, when I open the scheme editor, it immediately modifies the underlying scheme to re-add it _(even if it has no targets)_ - we can keep adding the `TestAction` to avoid unstable projects

```suggestion
            testAction: TestAction(
```

```diff
+++ b/projects/tuist/fixtures/ios_app_with_extensions/App.xcodeproj/xcshareddata/xcschemes/App.xcscheme
@@ -22,6 +22,14 @@
          </BuildActionEntry>
       </BuildActionEntries>
    </BuildAction>
+   <TestAction
+      buildConfiguration = "Debug"
+      selectedDebuggerIdentifier = "Xcode.DebuggerFoundation.Debugger.LLDB"
+      selectedLauncherIdentifier = "Xcode.DebuggerFoundation.Launcher.LLDB"
+      shouldUseLaunchSchemeArgsEnv = "YES">
+      <Testables>
+      </Testables>
+   </TestAction>
```

Reply:
If that's the case, I guess we should change the API to not get a nillable there (or change the internal logic to generate an empty test target in case of nil)? 🤔 
In both cases, I would do it in a separate PR